src/physics/em/detail/UrbanMscScatter.hh

```diff
@@ -0,0 +1,536 @@
+    static CELER_CONSTEXPR_FUNCTION Energy c_highland()
+    {
+        return Energy{13.6};
+    }
+
+    //// HELPER FUNCTIONS ////
+
+    // Sample the angle, cos(theta), of the multiple scattering
+    template<class Engine>
+    inline CELER_FUNCTION real_type sample_cos_theta(Engine&   rng,
+                                                     Energy    end_energy,
+                                                     real_type true_path,
+                                                     real_type limit_min);
+
+    // Sample the unit direction of the lateral displacement
+    template<class Engine>
+    inline CELER_FUNCTION Real3 dir_displacement(Engine&   rng,
+                                                 real_type phi) const;
+
+    // Sample consine(theta) with a large angle scattering
+    template<class Engine>
+    inline CELER_FUNCTION real_type simple_scattering(Engine&   rng,
+                                                      real_type xmean,
+                                                      real_type x2mean) const;
+
+    // Calculate the theta0 of the Highland formula
+    inline CELER_FUNCTION real_type compute_theta0(real_type true_path,
+                                                   Energy    end_energy) const;
+
+    // Calculate the correction on theta0 for positrons
+    inline CELER_FUNCTION real_type calc_correction(real_type tau) const;
+};
+
+//---------------------------------------------------------------------------//
+// INLINE DEFINITIONS
+//---------------------------------------------------------------------------//
+/*!
+ * Construct with shared and state data.
+ */
+CELER_FUNCTION
+UrbanMscScatter::UrbanMscScatter(const UrbanMscNativeRef& shared,
+                                 const ParticleTrackView& particle,
+                                 const Real3&             direction,
+                                 const PhysicsTrackView&  physics,
+                                 const MaterialView&      material,
+                                 const StepLimitResult&   input)
+    : inc_energy_(particle.energy())
+    , inc_direction_(direction)
+    , is_positron_(particle.particle_id() == shared.positron_id)
+    , rad_length_(material.radiation_length())
+    , mass_(shared.electron_mass.value())
+    , params_(shared.params)
+    , msc_(shared.msc_data[material.material_id()])
+    , helper_(shared, particle, physics, material)
+    , input_(input)
+{
+    CELER_EXPECT(particle.particle_id() == shared.electron_id
+                 || particle.particle_id() == shared.positron_id);
+
+    lambda_ = helper_.msc_mfp(inc_energy_);
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Sample the angular distribution and the lateral displacement by multiple
+ * scattering as well as convert the geometrical path length to the true path
+ * length based on G4VMultipleScattering::AlongStepDoIt of the Geant4 10.7
+ * release.
```

Comment:
And in G4UrbanMscModel::SampleScattering

Reply:
Added.